### DISCUSSION TITLE
Chatting about v1.0

### DISCUSSION THREAD

**Discussion by hay-kot** [mealie-recipes/mealie#645] [General]
Hey all it's been so cool to see how much everyone is enjoying using this thing I've put so much time into. Given where I was at in my programming knowledge when I started the project, I made a lot of poor design choices and wrote a lot of Javascript that isn't scalable. I've taken some time to assess the project and have determined that I really need to take a step back and re-look at how the Frontend is put together. Effectively I'm looking at a fullish rewrite of the Frontend. I've already started the process and am already seeing how the project will improve. 

Here's what to expect for the next few months on the stable branch

1. No new feature releases for Mealie. I'll be dedicating most of my time to the mealie-next branch and working on the rewrite.
2. I will still fix bugs as I can and time allows. Contributions will still be reviewed as well. 

The new version will be targeting v1.0 which means we're shooting for stability and a non-breaking API moving forward from their. No more backup/restoration process 🎉. We'll be moving to a 2 container setup with a Backend and a Frontend container with server side render which means faster loads speeds and more customizable sites. As of right now I already have a few really cool features working...

- Full Authentication Portal - Restrict your entire site to only authorized users
- Ability to run on a subpath 
- Open Signup - Can be disabled

More info will come soon, but in the mean-time feel free to provide feedback, leave kind words, or yell at me in the comments below.

I'm confident the changes I'm making will be all for the better 👍

P.S. if you're interested in helping out in the rewrite join the discord and get in touch. We're working with TypeScript and Nuxt.

**Comment by ghost** [comment]
Yeah I;m loving this. It's relazying.

- [ ] 

- [ ]

**Comment by k9sartech** [comment]
I've been out of the loop for a few months while waiting for the 5.0 update that was going to include printing and I see it's not only here; but there's already been two updates to it!  I checked out the print function and it's great.  Until I can project my recipes onto a cabinet door, I like using a print out copy of recipes.  This way I can make notes if it's a new recipe and I'm not getting my tablet or phone splattered.  You should see some of my recipes.  They've been used a lot and it shows!  

I saw a thread on Discord about your plans on different print options.  You mentioned about wanting it to scale like the King Arthur Page.  I looked at the link and that's exactly what I hope to be able to do with Mealie, too.  If I took the time to include notes and nutrition information in my recipe, I'm hoping I'd be able include those in the printout along with it's picture.  I'm a visual person.  I have a hard time visualizing a recipe without a picture attached.  That's why recipe books include pictures.  

Thank you for all the hard work you and everyone who's contributed to making Mealie what it is today.  It is really coming along and a great place to store cherished recipes.  Really looking forward to the different print functions to get implemented.

**Comment by tlskinneriv** [comment]
As requested, dropping my 2 cents from feature requests here about TOTP 2FA when authentication is reworked. Enjoying using the product and looking forward to contributing when I can.

**Comment by hay-kot** [comment]
# Update 8/25
I'm going to start posting periodic updates on the progress over in `mealie-next` branch here to try to keep others up to date. Here's what's new!

- The Two Containers are now running in demo capacity (mostly broken but running!) Lots of stuff doesn't work, so please don't file any issues. I'm aware of the problems. But you can take a look at [https://mealie.hay-kot.dev/demo/](https://mealie.hay-kot.dev/demo/)
- Side menu has been re-organized to group items together in a more concise way.
- Ingredient scaling is up and running! This includes an admin page for managing and creating units, a ton of default units/foods, and a robust recipe ingredient editor. The best part is, this feature is totally optional
- The Admin panel is now using a table-based approach to managing data which I think is more clear and consistent across the board. 
- Most of the work so far has been about migrating to typescript and Nuxt which has been such a joy with the new composition API. I'm finding I create fewer runtime errors and catch bugs faster which is great for everyone! 🎉

**Reply by zukedog** [reply]
> But you can take a look at https://mealie.hay-kot.dev/demo/

Seems like this link is broken. Just goes to a 404 Not Found.

**Reply by hay-kot** [reply]
@zukedog, ah, I moved everything over to a new domain and forgot to update. You can check it out [beta.mealie.io](https://beta.mealie.io)

**Reply by dzek69** [reply]
I cannot log in using the credentials displayed on top :(

**Reply by hay-kot** [reply]
@dzek69, that's cause I'm an idiot 😄 

Should be fixed now!

**Reply by DrSpaldo** [reply]
> @zukedog, ah, I moved everything over to a new domain and forgot to update. You can check it out [beta.mealie.io](https://beta.mealie.io)

Getting 404 for the new domain...

**Reply by fkastner** [reply]
> Getting 404 for the new domain...

It's https://demo.mealie.io now.

**Comment by bflac1** [comment]
Saw another post in the feature request discussion where you mentioned to put ideas here for authentication. Would be cool to see SSO integration to do single sign on with authentication applications like Authelia, Google Auth, etc. using something like OpenID Connect. I love this program and have been using it in my homelab to share recipes with the whole family. Thank you for great work and time you have put into this! Maybe one day ill be able to contribute more!

**Reply by joestump** [reply]
@bflac1 agreed. I am heavily biased towards self-hosted apps that work with my internal SSO provider (Authentik in my case). I can use LDAP with Authentik, but I prefer OIDC/OAuth2.

**Comment by r3df0chs** [comment]
Thanks for your work! Great piece of software!

**Comment by hay-kot** [comment]
New v1.0.0 docs are up at https://nightly.mealie.io/changelog/v1.0.0/

Largely they're the same, but I'll be updating them as I work through v1. Check the change-log for new features and improvements

**Comment by kreativmonkey** [comment]
Are there plans to integrate SSO with SAML or OpenID such as Authentik, Keycloak or Authelia?

**Reply by hay-kot** [reply]
Not at the moment. I don't deploy any of those services so the overhead for me to testing and configure is pretty high. I'm open to a PR if someone wants to work on an implementation.

**Comment by hay-kot** [comment]
The latest v1.0.0 demo now has an NLP model-driven ingredient parser https://beta.mealie.io/admin/parser

Huge step in making recipe scaling super easy.

**Reply by cashpw** [reply]
Love the ingredient parser! It's working like a charm.

There is a "Scaling" option when you click "Cook" in a recipe which doesn't appear to do anything. Are the ingredients meant to appear in this view?

![Home-Mealie](https://user-images.githubusercontent.com/56796112/140859129-3cadb069-effa-491a-8781-975048653cb2.png)

**Reply by hay-kot** [reply]
Just got that worked out this weekend, you have to link the ingredients to a step for them to show-up in the cook page. There's a some-what auto linking feature, but you still have to go through the dialog. 

<img width="904" alt="Screen Shot 2021-11-08 at 6 57 24 PM" src="https://user-images.githubusercontent.com/64056131/140859494-a66704ce-e4d2-42b0-92ce-926bfc232675.png">

**Comment by tyler351** [comment]
Just wanted to throw this out there real quick...
Came across your project about a month ago and finally had a chance to get is spun up. Such a cool thing to be able to run! 

AND saw your beta demo and immediately loved the rewrite! Keep up the great work!

**Reply by hay-kot** [reply]
Thanks for the kind words!

**Reply by RefineryX** [reply]
I also came here to say the same. I just installed Mealie and tbh, found it a little clunky and not the style I am used to. However, just found the BETA and this looks sooooooo much more improved! Its much simpler to use and gives me hope. 

Now I know what I'll be doing this weekend. Putting all my recipes into Mealie! Huge well done!

**Comment by hay-kot** [comment]
Just finished up work for the new meal-planner.

https://beta.mealie.io/meal-plan/planner

**Comment by RefineryX** [comment]
What is the reason v1.0.0 is split in two? (mealie-frontend and mealie-api).

Are both docker files required?

**Reply by hay-kot** [reply]
> We'll be moving to a 2 container setup with a Backend and a Frontend container with server side render which means faster loads speeds and more customizable sites.

That's pretty much it. Server side rendering allows ua to do cool things like generate social card previews for recipes and significantly speed up the site.

**Reply by Nelinski** [reply]
FYI having two separate dockers will break Unraid support as it doesn't (out the box) support docker compose.

**Reply by RefineryX** [reply]
Thanks. Whilst I love the new version, I am not a fan of having two docker images etc. Is there a minimal version of Mealie with none of the advanced stuff like API connections etc?

**Reply by hay-kot** [reply]
> Thanks. Whilst I love the new version, I am not a fan of having two docker images etc. Is there a minimal version of Mealie with none of the advanced stuff like API connections etc?

No that won't be an option. A sample docker-compose is provided in the beta documentation to show that it is quite easy to set-up [see here](https://nightly.mealie.io/documentation/getting-started/installation/sqlite/). Setting a single container requirement seems like an arbitrary limitation that would just hinder the development on the project. 

While it's possible to create a docker container that holds both the frontend and backend in one container, it sort of defeats the point of containers in the first place. If someone want's to develop that Dockerfile for that I'd be open to a PR, but it won't be the recommended way to install the application.

> FYI having two separate dockers will break Unraid support as it doesn't (out the box) support docker compose.

That is extremely unfortunate. I don't use Unraid or maintain the application template so I don't have much to offer here other than a quick google search showed that there are some work arounds for utilizing docker-compose on an Unraid box, but I'm unsure how effective those options are.

**Reply by Nelinski** [reply]
I've just tested this on Unraid and it's not too complicated to set up the two dockers separately. It all works as expected, so it could probably be tweaked to have two containers available via the community application page once this goes live/out of beta.

**Reply by hendrix04** [reply]
> While it's possible to create a docker container that holds both the frontend and backend in one container, it sort of defeats the point of containers in the first place.

While I agree with this from a scale / production perspective, for something like this, it makes way more sense to just use one container. Basically it is an over architected solution for no real gain.

**Reply by hay-kot** [reply]
> While I agree with this from a scale / production perspective, for something like this, it makes way more sense to just use one container. Basically it is an over architected solution for no real gain.

I disagree, but you're welcome to maintain a combined docker image. I'd be happy to provide a link to it in the documentation.

**Reply by RefineryX** [reply]
> > While it's possible to create a docker container that holds both the frontend and backend in one container, it sort of defeats the point of containers in the first place.
> 
> While I agree with this from a scale / production perspective, for something like this, it makes way more sense to just use one container. Basically it is an over architected solution for no real gain.

I agree. I hope someone takes up @hay-kot's offer to maintain a combined docker image. Currently I do not intend to install images which are split into multiple containers.

**Reply by hendrix04** [reply]
@RefineryX I am working on it. I am going a bit slow due to other priorities, but I hope to have something by the end of the week.

I will link to it here when I am done.

**Reply by RefineryX** [reply]
Nice work! Thanks @hendrix04

**Reply by BobWs** [reply]
Can test the v 1.0b? Where can I find the docker versions of it?

**Reply by hay-kot** [reply]
@BobWs - Checkout the nightly docs

https://nightly.mealie.io/documentation/getting-started/introduction/

**Reply by Nelinski** [reply]
@hendrix04 Did you get a working combined build? If so, got any instructions on what variables are needed to be set up etc? Would love to try it out.

**Comment by achurak** [comment]
I apologize in advance if it's dumb question and not specific to 1.0 (though as I understand food parser is a new 1.0 feature) cause I've just installed mealie for the first time and decided to start with 1.0 right away cause I need the shopping list feature. Why do all built-in ingredients have dashes instead of white spaces? And if there's a real reason behind that then why isn't the parser adjusted to recognize that?

Basically what it means right now is that I either have to manually select the existing "dashed" ingredients or create duplicates without dashes:

![image](https://user-images.githubusercontent.com/12438813/157481555-e951c0ec-1599-4879-87ae-314267056ce0.png)

 Also would be great if the parser understood that cherry tomato and cherry tomatoes is the same thing cause otherwise I'd have to have even more versions of the same ingredient.

**Comment by hay-kot** [comment]
> Why do all built-in ingredients have dashes instead of white spaces? And if there's a real reason behind that then why isn't the parser adjusted to recognize that?

This is a bug, the ingredients should be created without bugs. can you file an issues for this? 

> Also would be great if the parser understood that cherry tomato and cherry tomatoes is the same thing cause otherwise I'd have to have even more versions of the same ingredient.

I believe the parser should depluralize the ingredient before it get's processed. I'm using a pre-trained machine learning model to parse the ingredients, so unfortunately there's not much I can do outside of pre-process the strings. It should be understood that it's not silver bullet to automatically handle all cases, it's just there to help you with some of the easy cases - which will work much better when the bug you described is fixed. Thanks!

**Reply by achurak** [reply]
> This is a bug, the ingredients should be created without bugs. can you file an issues for this?

Done: https://github.com/hay-kot/mealie/issues/1025
 
> I believe the parser should depluralize the ingredient before it get's processed. I'm using a pre-trained machine learning model to parse the ingredients, so unfortunately there's not much I can do outside of pre-process the strings. It should be understood that it's not silver bullet to automatically handle all cases, it's just there to help you with some of the easy cases - which will work much better when the bug you described is fixed. Thanks!

I will test again, but I believe it offered to create "cherry tomatoes" while I've already added "cherry tomato" for another recipe. I understand you're not able to do anything about the pre-trained network, but am wondering if the ingredients can be depluralized/normalized before feeding them to the parser. I'd hate to have multiple varieties of the same ingredient just because of different spelling.

**Reply by hay-kot** [reply]
#1025 should be fixed with the latest build, thanks for catching that.

As for the other issue, I double checked and it looks like we're only singularizing units. I'm not sure how feasible it is to build something that will work universally. Mealie is translated into multiple languages and even just considering english there's enough edge cases to make me weary of trying to attempt something like this. 

Another solution may be to offer up near matches in the parser - I don't think this would be terrible all things considered, we could use some kind of fuzzy search and offer some near-match suggestions. Might be worth fleshing out a feature request on this to see if it's possible to implement without too much fuss.

**Comment by fivestones** [comment]
Are there instructions somewhere for how to install the v1.0 beta? I see the frontend-nightly and api-nightly tags on dockerhub but I'm not 100% sure how to use them.

**Reply by BobWs** [reply]
> Are there instructions somewhere for how to install the v1.0 beta? I see the frontend-nightly and api-nightly tags on dockerhub but I'm not 100% sure how to use them.

I have followed the instructions as described here https://nightly.mealie.io/documentation/getting-started/installation/sqlite/

**Comment by Nelinski** [comment]
What version of hhursev/recipe-scraper is currently being used in V1 (and how do I check)? How often is it updated?

**Reply by hay-kot** [reply]
You can check the pyproject.toml file or the poetry.lock file. Since we still haven't released we don't do regular updates, its mostly just updated when other dependencies are updated.

Once it's released we'll use dependabot to automatically get updates. If we miss an update you can always submit a PR to bump the version.

**Comment by madindehead** [comment]
I wanted to ask about the frontend and backend environment variables for the current nightly build. 

When I got to both links (https://nightly.mealie.io/mealie/documentation/getting-started/installation/frontend-config/ and https://nightly.mealie.io/mealie/documentation/getting-started/installation/backend-config/) I am seeing a 404 error. 

Is there anywhere else I can see these variables? Is this a known issue right now? 

Thanks

**Reply by Nelinski** [reply]
Share your config as it sounds like it's not set up correctly. All works fine for me.

**Reply by madindehead** [reply]
So this is just looking at those links - nothing to do with my Mealie config - so if there's a problem it's either my adblocker being too aggressive, or the page is broken.

**Reply by Nelinski** [reply]
Ah, yeah they aren't working for me either. These do though:
https://nightly.mealie.io/documentation/getting-started/installation/frontend-config/
https://nightly.mealie.io/documentation/getting-started/installation/backend-config/

**Reply by madindehead** [reply]
> Ah, yeah they aren't working for me either. These do though: https://nightly.mealie.io/documentation/getting-started/installation/frontend-config/ https://nightly.mealie.io/documentation/getting-started/installation/backend-config/

Excellent thanks.

**Comment by whirlfire** [comment]
Two questions regarding servings:
-Why is "Servings" not translated?
-Why is the servings-calculation not like e.g. in the Tandoori-Software? just multiplying the first number of each ingridient element (i guess). That should work like, nearly always without any compicated stuff.
The way it is now, it would be necessary to adjust each imported recipe and its ingridients one by one.

Apart from that, its a really cool project. Thank you for your effort!

**Reply by hay-kot** [reply]
> Why is "Servings" not translated?

There are many missing translations in the application. I'm only one person and can only do so much. You're welcome to help out and add more text to the translation system. 

> Why is the servings-calculation not like e.g. in the Tandoori-Software? 

That's not how Tandoor works. It will automatically parse you ingredients into units, foods, and notes. When I tried it out I had a poor experience overall and it would often delete or mess up the ingredients. As such, I've opted for an opt-int approach that allows users to choose if they're going to go through the hassle of parsing out the units, foods, amounts, and notes. I often don't use this feature at all, so having this option is ideal for myself. 

> Just multiplying the first number of each ingridient element (i guess). That should work like, nearly always without any compicated stuff.

This is a widely over simplification of what is a very difficult problem to solve. As many users can attest, there many recipes that do-not ad hear to any common patterns for text and trying to do some kind of pattern matching to find a number would yield widely inaccurate results and result in lots of Github issue and edge cases to program for. There is a length discussion on the implementation and ideas around scaling ingredients. See Here https://github.com/hay-kot/mealie/discussions/694

**Comment by mostlyvirtual** [comment]
@hay-kot thank you so much for all the work you've put in. I wonder what was your main reason for making the decission to make the site login only, and essentially not have a public recipe book and/or a shared recipes section where people cand all put their recipes together? At one point, in issue #597 you even said that you won't be implementing this, so I'm curious why the change of heart.

I've been waiting for v1 for some awesome features that you've implemented, but it now breaks the use case I wanted to use this for unfortunately.

My apologies if this was discussed somewhere else, but I couldn't find it with the keywords I used to search.

**Reply by hay-kot** [reply]
This functionality was requested multiple times a week until it was implemented. Basically, I was worn down over time. It also was a somewhat required trade-off for the multi-tenancy support and proper isolation of group data. I agree that there is still a need for an optional public facing way to view a groups recipes. I've outlined those here

-  https://github.com/hay-kot/mealie/issues/1326

I think that should fulfill most use cases but it's not slated until beta 5 so it's a ways away. If you or anyone else wants to contribute I'm willing to accept a PR on that functionality, or you can hire me to freelance on the project and implement it early. Otherwise, you can wait it out until it get's implemented.

**Comment by nealhamiltonjr** [comment]
Not only would this be awesome for food but also for brewing or distilling recipes. I tried to scrape this url but it didn't work.

https://www.homebrewersassociation.org/homebrew-recipe/rushmore-ipa-extract/

**Comment by madindehead** [comment]
I have to say, I find the Meal Planner in the latest version much less intuitive than the 0.5.x versions. I have made a rule for a meal plan, but it only populated three days. When I try to add a random dinner, it says I have no recipes that match my rules, despite having recipes in the categories that I've selected for dinners. 

Seems when I press the 'Random Dinner' button, I get a 404 error after the POST request. Related to the XHR.

**Reply by thuesen** [reply]
@madindehead  did you ever find a solution to this?

I can't figure how to add a random dinner on meal plan either.  If I add a random Breakfast or Lunch, it selects from all my recipes it seems, even dinner recipes.

**Comment by bartmans** [comment]
First of all thanks for all the hard work! Mealie works intuitive and fast.

I really don't have that much features I find missing. For me, most important is stability. If I commit to using Mealie instead of my trusted stack of printed out recipes, than it must be stable. Migrating from 0.5 tot stable-3 scared me a bit. So (commitment to) backwards compatibility for recipes, categories and labels is high on my list.
